src/ray/core_worker/transport/out_of_order_actor_submit_queue.cc

```diff
@@ -0,0 +1,85 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/transport/out_of_order_actor_submit_queue.h"
+
+namespace ray {
+namespace core {
+OutofOrderActorSubmitQueue::OutofOrderActorSubmitQueue(ActorID actor_id)
+    : actor_id(actor_id) {}
+
+bool OutofOrderActorSubmitQueue::Emplace(uint64_t position, TaskSpecification spec) {
+  return requests.emplace(position, std::make_pair(spec, /*dependency_resolved*/ false))
+      .second;
+}
+
+bool OutofOrderActorSubmitQueue::Contains(uint64_t position) const {
+  return requests.find(position) != requests.end();
+}
+
+const std::pair<TaskSpecification, bool> &OutofOrderActorSubmitQueue::Get(
+    uint64_t position) const {
+  auto it = requests.find(position);
+  RAY_CHECK(it != requests.end());
+  return it->second;
+}
+
+void OutofOrderActorSubmitQueue::MarkDependencyFailed(uint64_t position) {
+  requests.erase(position);
+}
+
+void OutofOrderActorSubmitQueue::MarkDependencyResolved(uint64_t position) {
+  auto it = requests.find(position);
+  RAY_CHECK(it != requests.end());
+  it->second.second = true;
+}
+
+std::vector<TaskID> OutofOrderActorSubmitQueue::ClearAllTasks() {
+  std::vector<TaskID> task_ids;
+  for (auto &[pos, spec] : requests) {
+    task_ids.push_back(spec.first.TaskId());
+  }
+  requests.clear();
+  return task_ids;
+}
+
+absl::optional<std::pair<TaskSpecification, bool>>
+OutofOrderActorSubmitQueue::PopNextTaskToSend() {
+  for (auto it = requests.begin(); it != requests.end(); it++) {
+    if (/*dependencies_resolved*/ it->second.second) {
```

Comment:
Maybe add a TODO to optimize this so it's not O(n^2) in the worst case (but seems unlikely to be an issue).

Reply:
I think we can easily do that by using 2 queues? (one for dep_resolved queue, and the other for non resolved queue)? Why don't we just do this here?